File: ruby/lib/epostak/resources/webhooks.rb
# frozen_string_literal: true

require "erb"
require "openssl"

module EPostak
  # Verify an inbound webhook signature. Uses constant-time comparison to prevent
  # timing attacks. Rejects payloads older than +max_age_ms+ milliseconds to
  # prevent replay attacks.
  #
  # Signature format: +t=<unix_ms>,v1=<hex_hmac_sha512>+
  #
  # @param payload [String] The raw webhook request body
  # @param signature [String] The value of the +X-Epostak-Signature+ header
  # @param secret [String] The webhook signing secret from webhook creation
  # @param max_age_ms [Integer] Maximum acceptable age in milliseconds (default 300_000 = 5 min)
  # @return [Boolean] +true+ if the signature is valid and within the replay window
  #
  # @example
  #   valid = EPostak::Webhook.verify_signature(
  #     payload: request.body.read,
  #     signature: request.headers["X-Epostak-Signature"],
  #     secret: ENV["WEBHOOK_SECRET"]
  #   )
  #   return head :unauthorized unless valid
  module Webhook
    MAX_AGE_MS = 300_000

    # @see EPostak::Webhook
    def self.verify_signature(payload:, signature:, secret:, max_age_ms: MAX_AGE_MS)
      t_part = nil
      v_part = nil
      signature.split(",").each do |segment|
        t_part = segment[2..] if segment.start_with?("t=")
        v_part = segment[3..] if segment.start_with?("v1=")
      end
      return false if t_part.nil? || v_part.nil?

      timestamp = Integer(t_part, 10, exception: false)
      return false if timestamp.nil?

      age = (Time.now.to_f * 1000).to_i - timestamp
      return false if age < 0 || age > max_age_ms

      message = "#{timestamp}.#{payload}"
      computed = OpenSSL::HMAC.hexdigest("SHA512", secret, message)

      OpenSSL.secure_compare(computed, v_part.downcase)
    rescue StandardError
      false
    end
  end

  module Resources
    # Resource for managing webhook subscriptions and the pull queue.
    #
    # Webhooks notify your server about document events (sent, received, validated).
    # Choose between push webhooks (server receives HTTPS POST) or the pull queue
    # (your code polls for events).
    #
    # @example Create a push webhook
    #   webhook = client.webhooks.create(
    #     url: "https://example.com/webhooks/epostak",
    #     events: ["document.received", "document.sent"]
    #   )
    #   # Store webhook["secret"] for HMAC verification
    class Webhooks
      # @return [Resources::WebhookQueue] Sub-resource for the pull queue (polling-based event consumption)
      attr_reader :queue

      # @param http [EPostak::HttpClient] Internal HTTP client
      def initialize(http)
        @http  = http
        @queue = WebhookQueue.new(http)
      end

      # Create a new webhook subscription. Returns the HMAC-SHA256 signing secret
      # which is only available at creation time -- store it securely.
      #
      # @param url [String] HTTPS URL to receive webhook POST requests. HTTPS is required; +http://+ URLs are rejected.
      # @param events [Array<String>, nil] Event types to subscribe to (nil = all events).
      #   Available events (7): +document.created+, +document.sent+, +document.received+,
      #   +document.validated+, +document.delivered+, +document.rejected+,
      #   +document.response_received+.
      # @return [Hash] Webhook details { "id", "url", "events", "secret", "isActive", "createdAt" }
      #   — +secret+ is shown only once; store it securely.
      #
      # @param idempotency_key [String, nil] Optional client-chosen
      #   `Idempotency-Key` header. Replays of the same key surface as
      #   `EPostak::Error` with `code == "idempotency_conflict"`.
      #
      # @example
      #   webhook = client.webhooks.create(
      #     url: "https://example.com/webhooks",
      #     events: ["document.received"]
      #   )
      #   puts webhook["secret"] # => Store this securely!
      def create(url:, events: nil, idempotency_key: nil)
        body = { url: url }
        body[:events] = events if events
        @http.request(:post, "/webhooks", body: body, idempotency_key: idempotency_key)
      end

      # List all webhook subscriptions for the current account.
      #
      # @return [Hash] Response with "data" array of webhook objects
      #
      # @example
      #   response = client.webhooks.list
      #   response["data"].each { |w| puts "#{w['url']} (active: #{w['isActive']})" }
      def list
        @http.request(:get, "/webhooks")
      end

      # Get a webhook subscription by ID, including recent delivery history.
      # Use the delivery history to debug failed webhook deliveries.
      #
      # @param id [String] Webhook UUID
      # @return [Hash] Webhook details with "deliveries" array
      #
      # @example
      #   webhook = client.webhooks.get("webhook-uuid")
      #   failed = webhook["deliveries"].select { |d| d["status"] == "failed" }
      def get(id)
        @http.request(:get, "/webhooks/#{encode(id)}")
      end

      # Update a webhook subscription. Use this to change the URL, event filter,
      # or pause/resume the webhook.
      #
      # @param id [String] Webhook UUID
      # @param params [Hash] Fields to update (url, events, isActive)
      # @return [Hash] The updated webhook
      #
      # @example Pause a webhook
      #   client.webhooks.update("webhook-uuid", isActive: false)
      #
      # @example Change URL and events
      #   client.webhooks.update("webhook-uuid",
      #     url: "https://new-url.com/webhooks",
      #     events: ["document.received", "document.validated"]
      #   )
      def update(id, **params)
        @http.request(:patch, "/webhooks/#{encode(id)}", body: params)
      end

      # Delete a webhook subscription. Stops all future deliveries.
      #
      # Returns HTTP 204 No Content on success; the SDK returns +nil+.
      #
      # @param id [String] Webhook UUID to delete
      # @return [nil] Nothing on success (HTTP 204 No Content)
      #
      # @example
      #   client.webhooks.delete("webhook-uuid") # => nil
      def delete(id)
        @http.request(:delete, "/webhooks/#{encode(id)}")
      end

      # Send a test event to a webhook endpoint.
      #
      # @param id [String] Webhook UUID to test
      # @param event [String, nil] Event type to simulate (e.g. "document.created"). Nil uses server default.
      # @return [Hash] Test result with "success", "statusCode", "responseTime", "webhookId", "event", and optional "error"
      #
      # @example
      #   result = client.webhooks.test("webhook-uuid", event: "document.received")
      #   puts result["success"] ? "OK" : result["error"]
      def test(id, event: nil)
        body = {}
        body[:event] = event if event
        @http.request(:post, "/webhooks/#{encode(id)}/test", body: body)
      end

      # Get paginated delivery history for a webhook.
      #
      # @param id [String] Webhook UUID
      # @param limit [Integer, nil] Max deliveries to return (1-100, default 20)
      # @param offset [Integer, nil] Number of deliveries to skip (default 0)
      # @param status [String, nil] Filter by status: "SUCCESS", "FAILED", "PENDING", "RETRYING"
      # @param event [String, nil] Filter by event type
      # @return [Hash] Paginated response with "deliveries", "total", "limit", "offset"
      #
      # @example
      #   result = client.webhooks.deliveries("webhook-uuid", status: "FAILED", limit: 50)
      #   result["deliveries"].each { |d| puts "#{d['event']}: #{d['status']}" }
      def deliveries(id, **params)
        @http.request(:get, "/webhooks/#{encode(id)}/deliveries", query: params)
      end

      # Rotate a webhook's HMAC-SHA256 signing secret. Issues a fresh secret
      # and invalidates the previous one immediately. The new secret is
      # returned ONCE — store it right away; there is no way to retrieve it
      # later. In-flight deliveries signed with the old secret will no longer
      # verify on the receiving side. Non-destructive alternative to
      # delete+recreate when a secret leaks.
      #
      # @param id [String] Webhook UUID whose secret to rotate
      # @return [Hash] { "id" => ..., "secret" => ..., "message" => ... } — secret only shown once
      #
      # @example
      #   res = client.webhooks.rotate_secret("webhook-uuid")
      #   ENV["EPOSTAK_WEBHOOK_SECRET"] = res["secret"]
      def rotate_secret(id)
        @http.request(:post, "/webhooks/#{encode(id)}/rotate-secret")
      end

      private

      def encode(value)
        ERB::Util.url_encode(value)
      end
    end
  end
end

File: python/src/epostak/resources/webhooks.py
"""Webhooks resource -- manage webhook subscriptions and pull queue.

Provides :class:`WebhooksResource` for creating, updating, and deleting push
webhook subscriptions, and :class:`WebhookQueueResource` (accessible via
``client.webhooks.queue``) for polling-based event consumption.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, TYPE_CHECKING
from urllib.parse import quote

if TYPE_CHECKING:
    from epostak.types import (
        Webhook,
        WebhookDetail,
        WebhookQueueAllResponse,
        WebhookQueueResponse,
        WebhookWithDeliveries,
    )

from epostak.resources.documents import _BaseResource, _build_query

import httpx


class WebhookQueueResource(_BaseResource):
    """Webhook pull queue for polling-based event consumption."""

    def pull(
        self,
        limit: Optional[int] = None,
        event_type: Optional[str] = None,
    ) -> WebhookQueueResponse:
        """Fetch pending webhook events from the pull queue.

        Args:
            limit: Max items to return, 1-100 (default 20).
            event_type: Filter by event type, e.g. ``"document.received"``.

        Returns:
            Dict with ``items`` (list of queue events) and ``has_more`` (bool).

        Example::

            queue = client.webhooks.queue.pull(limit=50)
            for item in queue["items"]:
                print(item["type"], item["payload"])
        """
        params = _build_query({"limit": limit, "event_type": event_type})
        return self._request("GET", "/webhook-queue", params=params)

    def ack(self, event_id: str) -> None:
        """Acknowledge a single webhook event (removes it from the queue).

        Args:
            event_id: The event UUID to acknowledge.

        Returns:
            None (HTTP 204).

        Example::

            client.webhooks.queue.ack("event-uuid")
        """
        self._request("DELETE", f"/webhook-queue/{quote(event_id, safe='')}")

    def batch_ack(self, event_ids: List[str]) -> None:
        """Batch acknowledge webhook events (removes them from the queue).

        Args:
            event_ids: List of event UUIDs to acknowledge.

        Returns:
            None (HTTP 204).

        Example::

            ids = [item["id"] for item in queue["items"]]
            client.webhooks.queue.batch_ack(ids)
        """
        self._request("POST", "/webhook-queue/batch-ack", json={"event_ids": event_ids})

    def pull_all(
        self,
        limit: Optional[int] = None,
        since: Optional[str] = None,
    ) -> WebhookQueueAllResponse:
        """Fetch events across all firms (integrator only).

        Args:
            limit: Max items, 1-500 (default 100).
            since: ISO 8601 timestamp -- only return events created after this time.

        Returns:
            Dict with ``events`` (each including ``firm_id``) and ``count``.

        Example::

            queue = client.webhooks.queue.pull_all(limit=200)
            for event in queue["events"]:
                print(event["firm_id"], event["event"])
        """
        params = _build_query({"limit": limit, "since": since})
        return self._request("GET", "/webhook-queue/all", params=params)

    def batch_ack_all(self, event_ids: List[str]) -> Dict[str, int]:
        """Cross-firm batch acknowledge (integrator only).

        Args:
            event_ids: List of event UUIDs from any firm.

        Returns:
            Dict with ``acknowledged`` count.

        Example::

            ids = [e["event_id"] for e in queue["events"]]
            result = client.webhooks.queue.batch_ack_all(ids)
            print(result["acknowledged"])
        """
        return self._request("POST", "/webhook-queue/all/batch-ack", json={"event_ids": event_ids})


class WebhooksResource(_BaseResource):
    """Manage webhook subscriptions and access the pull queue."""

    queue: WebhookQueueResource
    """Access the webhook pull queue for polling-based consumption."""

    def __init__(
        self,
        client: httpx.Client,
        base_url: str,
        api_key: str,
        firm_id: Optional[str],
        *,
        max_retries: int = 3,
    ) -> None:
        super().__init__(client, base_url, api_key, firm_id, max_retries=max_retries)
        self.queue = WebhookQueueResource(client, base_url, api_key, firm_id, max_retries=max_retries)

    def create(
        self,
        url: str,
        events: Optional[List[str]] = None,
    ) -> WebhookDetail:
        """Register a webhook endpoint.

        Args:
            url: The HTTPS URL to deliver webhook events to.
            events: Optional list of event types to subscribe to.  If omitted, all
                events are delivered.  Valid types: ``"document.created"``,
                ``"document.sent"``, ``"document.received"``, ``"document.validated"``.

        Returns:
            Webhook detail including the HMAC-SHA256 signing ``secret`` -- store it
            securely for verifying webhook signatures.

        Example::

            webhook = client.webhooks.create(
                url="https://example.com/webhook",
                events=["document.received", "document.sent"],
            )
            print(webhook["id"], webhook["secret"])
        """
        body: Dict[str, Any] = {"url": url}
        if events is not None:
            body["events"] = events
        return self._request("POST", "/webhooks", json=body)

    def list(self) -> List[Webhook]:
        """List all webhook subscriptions.

        Returns:
            List of webhook dicts with ``id``, ``url``, ``events``, ``isActive``,
            and ``createdAt``.

        Example::

            webhooks = client.webhooks.list()
            for wh in webhooks:
                print(wh["url"], wh["isActive"])
        """
        res = self._request("GET", "/webhooks")
        return res.get("data", res) if isinstance(res, dict) else res

    def get(self, id: str) -> WebhookWithDeliveries:
        """Get webhook detail with recent delivery history.

        Args:
            id: Webhook UUID.

        Returns:
            Webhook dict including ``deliveries`` list with recent delivery attempts.

        Example::

            detail = client.webhooks.get("webhook-uuid")
            for d in detail["deliveries"]:
                print(d["event"], d["status"], d["responseStatus"])
        """
        return self._request("GET", f"/webhooks/{quote(id, safe='')}")

    def update(
        self,
        id: str,
        url: Optional[str] = None,
        events: Optional[List[str]] = None,
        is_active: Optional[bool] = None,
    ) -> Webhook:
        """Update a webhook subscription.

        All parameters besides ``id`` are optional -- only pass the fields you want
        to change.

        Args:
            id: Webhook UUID.
            url: New delivery URL.
            events: New list of event types to subscribe to.
            is_active: Set to ``False`` to pause delivery, ``True`` to resume.

        Returns:
            The updated webhook object.

        Example::

            client.webhooks.update("webhook-uuid", is_active=False)
        """
        body: Dict[str, Any] = {}
        if url is not None:
            body["url"] = url
        if events is not None:
            body["events"] = events
        if is_active is not None:
            body["isActive"] = is_active
        return self._request("PATCH", f"/webhooks/{quote(id, safe='')}", json=body)

    def delete(self, id: str) -> None:
        """Delete a webhook subscription.

        Args:
            id: Webhook UUID.

        Returns:
            None (HTTP 204).

        Example::

            client.webhooks.delete("webhook-uuid")
        """
        self._request("DELETE", f"/webhooks/{quote(id, safe='')}")

    def test(self, id: str, event: Optional[str] = None) -> Dict[str, Any]:
        """Send a test event to a webhook endpoint.

        Args:
            id: Webhook UUID to test.
            event: Event type to simulate, e.g. ``"document.created"``.
                If omitted, the server picks a default event.

        Returns:
            Dict with ``success`` (bool), ``statusCode`` (int or None),
            ``responseTime`` (int), ``webhookId``, ``event``, and optional ``error``.

        Example::

            result = client.webhooks.test("webhook-uuid", event="document.received")
            print(result["success"], result["responseTime"])
        """
        body: Dict[str, Any] = {}
        if event is not None:
            body["event"] = event
        return self._request("POST", f"/webhooks/{quote(id, safe='')}/test", json=body)

    def deliveries(
        self,
        id: str,
        *,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
        status: Optional[str] = None,
        event: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Get paginated delivery history for a webhook.

        Args:
            id: Webhook UUID.
            limit: Max deliveries to return, 1-100 (default 20).
            offset: Number of deliveries to skip (default 0).
            status: Filter by status: ``"SUCCESS"``, ``"FAILED"``, ``"PENDING"``, ``"RETRYING"``.
            event: Filter by event type, e.g. ``"document.received"``.

        Returns:
            Dict with ``deliveries`` list, ``total``, ``limit``, and ``offset``.

        Example::

            result = client.webhooks.deliveries("webhook-uuid", status="FAILED", limit=50)
            for d in result["deliveries"]:
                print(d["event"], d["status"], d["attempts"])
        """
        params = _build_query({"limit": limit, "offset": offset, "status": status, "event": event})
        return self._request("GET", f"/webhooks/{quote(id, safe='')}/deliveries", params=params)

    def rotate_secret(self, id: str) -> Dict[str, Any]:
        """Rotate a webhook's HMAC-SHA256 signing secret.

        Issues a fresh signing secret and invalidates the previous one
        immediately. The new secret is returned ONCE -- store it right
        away. In-flight deliveries signed with the old secret will no
        longer verify on the receiving side. Non-destructive alternative
        to deleting and recreating the webhook when a secret leaks.

        Args:
            id: Webhook UUID whose secret to rotate.

        Returns:
            Dict with ``id``, ``secret`` (only shown once), and ``message``.

        Example::

            res = client.webhooks.rotate_secret("webhook-uuid")
            save_to_secrets_manager(res["secret"])
        """
        return self._request("POST", f"/webhooks/{quote(id, safe='')}/rotate-secret")
